geth/peers/topicpool.go

```diff
@@ -0,0 +1,234 @@
+package peers
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common/mclock"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/p2p/discover"
+	"github.com/ethereum/go-ethereum/p2p/discv5"
+	"github.com/status-im/status-go/geth/params"
+)
+
+// NewTopicPool returns instance of TopicPool
+func NewTopicPool(topic discv5.Topic, limits params.Limits, slowSync, fastSync time.Duration) *TopicPool {
+	return &TopicPool{
+		topic:    topic,
+		limits:   limits,
+		slowSync: slowSync,
+		fastSync: fastSync,
+		peers:    map[discv5.NodeID]*peerInfo{},
+	}
+}
+
+// TopicPool manages peers for topic.
+type TopicPool struct {
+	topic    discv5.Topic
+	limits   params.Limits
+	slowSync time.Duration
+	fastSync time.Duration
+
+	quit    chan struct{}
+	running int32
+
+	mu         sync.RWMutex
+	discvWG    sync.WaitGroup
+	consumerWG sync.WaitGroup
+	connected  int
+	peers      map[discv5.NodeID]*peerInfo
+	period     chan time.Duration
+
+	cache *Cache
+}
+
+// SearchRunning returns true if search is running
+func (t *TopicPool) SearchRunning() bool {
+	return atomic.LoadInt32(&t.running) == 1
+}
+
+// MaxReached returns true if we connected with max number of peers.
+func (t *TopicPool) MaxReached() bool {
+	t.mu.RLock()
+	defer t.mu.RUnlock()
+	return t.connected == t.limits[1]
+}
+
+// ConfirmAdded called when peer was added by p2p Server
+func (t *TopicPool) ConfirmAdded(server *p2p.Server, nodeID discover.NodeID) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	// inbound connection
+	peer, exist := t.peers[discv5.NodeID(nodeID)]
+	if !exist {
+		return
+	}
+	// established connection means that the node is a viable candidate for a connection and can be cached
+	if t.cache != nil {
+		if err := t.cache.AddPeer(peer.node, t.topic); err != nil {
+			log.Error("failed to persist a peer", "error", err)
+		}
+	}
+	// when max limit is reached drop every peer after
+	if t.connected == t.limits[1] {
+		log.Debug("max limit is reached drop the peer", "ID", nodeID, "topic", t.topic)
+		t.removePeer(server, peer)
+		return
+	}
+	// don't count same peer twice
+	if !peer.connected {
+		log.Debug("marking as connected", "ID", nodeID)
+		peer.connected = true
+		t.connected++
+	}
+	if t.SearchRunning() && t.connected == t.limits[0] {
+		t.period <- t.slowSync
+	}
+	return
+}
+
+// ConfirmDropped called when server receives drop event
+func (t *TopicPool) ConfirmDropped(server *p2p.Server, nodeID discover.NodeID, reason string) (new bool) {
+	t.mu.Lock()
+	defer t.mu.Unlock()
+	// either inbound or connected from another topic
+	peer, exist := t.peers[discv5.NodeID(nodeID)]
+	if !exist {
+		return false
+	}
+	log.Debug("disconnect reason", "peer", nodeID, "reason", reason)
+	// if requested - we don't need to remove peer from cache and look for a replacement
+	if reason == p2p.DiscRequested.Error() {
```

Comment:
if we are removing a peer with server.RemovePeer - that same peer will appear in this event with DiscRequested.
i reworked it in https://github.com/status-im/status-go/pull/801 though because we can get the same error if remote peer will drop us